File: intersection-of-two-linked-lists.cpp
/*
Initialize two pointers (a and b) at the heads of the linked lists. 
Traverse both lists simultaneously, resetting pointers to the other list's head when they reach the end.
If there is an intersection, the pointers will meet at the intersection node; otherwise, they will reach the end of both lists. Return the intersection node if found or nullptr if no intersection.
Time Complexity: O(M + N), where M and N are the lengths of the two linked lists.
Space Complexity: O(1), as it uses a constant amount of extra space regardless of the input size.*/
#include<bits/stdc++.h>
using namespace std;
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if(!headA ||!headB)return nullptr;
        ListNode *a=headA;
        ListNode *b=headB;
        while(a!=b){
            a=(a)?a->next:headB;
            b=(b)?b->next:headA;
            if(a==b)return a;
        }
        return a;
    }
};

File: intersection of two linked lists.py
#tc: O(m + n)
#sc: O(1)
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        if headA and headB is None: #base case
            return None

        lenA = 0 #initializing 
        lenB = 0
        curr = headA
        while curr is not None: #calculating length A
            curr = curr.next
            lenA += 1

        curr = headB #using the same pointer to calculate length B
        while curr is not None:
            curr = curr.next
            lenB += 1
        
        while lenA > lenB: #if the upper list is long
            headA = headA.next
            lenA -= 1

        while lenB > lenA: #if the below list is long
            headB = headB.next
            lenB -= 1

        while headA != headB: #till both the nodes intersect
            headA = headA.next
            headB = headB.next
        
        return headA
        

        
        
